src/winwatch/winwatch_stubs.c

```diff
@@ -0,0 +1,207 @@
+#include <stdio.h>
+#include <caml/mlvalues.h>
+#include <caml/alloc.h>
+#include <caml/fail.h>
+#include <caml/callback.h>
+#include <caml/osdeps.h>
+#include <caml/memory.h>
+#include <caml/threads.h>
+#include <caml/unixsupport.h>
+#include <caml/custom.h>
+
+#ifdef _WIN32
+
+#include <windows.h>
+
+#define BUFF_SIZE 1024
+
+typedef struct
+{
+    char *buffer;
+    HANDLE handle;
+    OVERLAPPED *overlapped;
+    value func;
+} dune_t;
+
+static void winwatch_watch(dune_t *t)
+{
+    memset(t->overlapped, 0, sizeof(OVERLAPPED));
+
+    BOOL watch_path = ReadDirectoryChangesW(
+            t->handle, t->buffer, BUFF_SIZE, TRUE,
+            FILE_NOTIFY_CHANGE_FILE_NAME  |
+            FILE_NOTIFY_CHANGE_DIR_NAME   |
+            FILE_NOTIFY_CHANGE_LAST_WRITE,
+            NULL, t->overlapped, NULL);
+}
+
+value winwatch_iocp_create(value v_unit)
+{
+    CAMLparam1(v_unit);
+    HANDLE port = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 1);
+    if (port == NULL)
+      caml_failwith("CreateIoCompletionPort failed");
+    CAMLreturn(caml_copy_nativeint((intnat)port));
+}
+
+value winwatch_iocp_run(value v_iocp)
+{
+    CAMLparam1(v_iocp);
+    CAMLlocal2(v_file_name, v_action);
+    HANDLE iocp = (HANDLE)Nativeint_val(v_iocp);
+    DWORD num_bytes, name_len;
+    WCHAR* file_name;
+    OVERLAPPED *overlapped;
+    FILE_NOTIFY_INFORMATION *event;
+    dune_t *t;
+
+    while (TRUE)
+    {
+        caml_release_runtime_system();
+        BOOL ok = GetQueuedCompletionStatus(iocp, &num_bytes, (PULONG_PTR)&t, &overlapped, 100);
+        caml_acquire_runtime_system();
+
+        if (ok == FALSE && overlapped == NULL) /* Timeout */
+            continue;
+
+        if (ok == FALSE)
+            /* Ignore errors */
+            continue;
+
+        event = (FILE_NOTIFY_INFORMATION*)t->buffer;
+
+        for (;;)
+        {
+            name_len = event->FileNameLength / sizeof(WCHAR);
+
+            switch (event->Action)
+            {
+                case FILE_ACTION_ADDED:
+                    v_action = Val_int(0);
+                    break;
+                case FILE_ACTION_REMOVED:
+                    v_action = Val_int(1);
+                    break;
+                case FILE_ACTION_MODIFIED:
+                    v_action = Val_int(2);
+                    break;
+                case FILE_ACTION_RENAMED_OLD_NAME:
+                    v_action = Val_int(3);
+                    break;
+                case FILE_ACTION_RENAMED_NEW_NAME:
+                    v_action = Val_int(4);
+                    break;
+            }
+
+            file_name = malloc(sizeof(WCHAR) * (name_len + 1));
```

Comment:
`file_name` is not freed (it can be right after it's been copied in `v_file_name`)